### PR TITLE
Spring security 2

### PR DESCRIPTION
**Pull request by layzimayner**

### CODE REVIEW

File: src/main/java/com/example/demo/model/User.java [hunk 1]
```diff
@@ -26,4 +33,41 @@ public class User {
...
+    }
+
+    @Override
+    public boolean isCredentialsNonExpired() {
+        return true;
+    }
+
+    @Override
+    public boolean isEnabled() {
+        return true;
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/model/User.java]:
```suggestion
        return !isDeleted;
```
Let’s  implement soft-deleted

File: src/main/java/com/example/demo/model/User.java [hunk 2]
```diff
@@ -5,15 +5,22 @@
...
 import lombok.Setter;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.userdetails.UserDetails;
 
 @Entity
 @Getter
 @Setter
 @Table(name = "users")
-public class User {
+public class User implements UserDetails {
```
Comment (liliia-ponomarenko) [inline src/main/java/com/example/demo/model/User.java:27]: Let’s  fix the registration flow 
1) encode password 
2) give USER default role 
3) fix the registration path

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let’s improve your solution ;)

File: src/main/java/com/example/demo/model/User.java (hunk 1 above)
Comment (layzimayner) [reply src/main/java/com/example/demo/model/User.java]: if i haven't endpoint wich delete user i still need soft delete?

File: src/main/java/com/example/demo/model/User.java [hunk 3]
```diff
@@ -26,4 +37,43 @@ public class User {
     @Column(nullable = false)
     private String lastName;
     private String shippingAddress;
+    @ManyToMany
+    @JoinTable(
+            name = "user_roles",
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/User.java]:
```suggestion
            name = "users_roles",
```

File: src/main/java/com/example/demo/security/CustomUserDetailsService.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+@Service
+@RequiredArgsConstructor
+public class CustomUserDetailsService implements UserDetailsService {
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetails loadUserByUsername(String email)
+            throws UsernameNotFoundException {
+        return userRepository.findByEmail(email)
+                .orElseThrow(() -> new UsernameNotFoundException("Can't find user by email"));
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/security/CustomUserDetailsService.java]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Can't find user by email"));
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
just two comments

File: src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      author: coder
+      preConditions:
+        onFail: MARK_RAN
+        conditions:
+          - not:
+              tableExists:
+                tableName: user_roles
+      changes:
+        - createTable:
+            tableName: user_roles
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml]:
```suggestion
            tableName: users_roles
```

File: src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml [hunk 6]
```diff
@@ -0,0 +1,39 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-user-roles-table
+      author: coder
+      preConditions:
+        onFail: MARK_RAN
+        conditions:
+          - not:
+              tableExists:
+                tableName: user_roles
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml]: what for?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

File: src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml (hunk 6 above)
Comment (layzimayner) [reply src/main/resources/db/changelog/changes/04-crate-user-role-table.yaml]: the project did not build because the table already existed

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😊

File: src/main/java/com/example/demo/model/User.java (hunk 1 above)
Comment (liliia-ponomarenko) [reply src/main/java/com/example/demo/model/User.java]: yes, you are setting this up for an account deactivation feature that may be implemented in the future

**Pull request merged**